frontend/src/components/Tree/Tree.jsx

```diff
@@ -0,0 +1,138 @@
+import React, { useState, useRef, useEffect } from "react";
+import styles from "./Tree.module.css";
+import { select, hierarchy, tree, linkHorizontal } from "d3";
+import ResizeObserver from "resize-observer-polyfill";
+import "./Tree.module.css";
+
+const useResizeObserver = (ref) => {
+  const [dimensions, setDimensions] = useState(null);
+  useEffect(() => {
+    const observeTarget = ref.current;
+    const resizeObserver = new ResizeObserver((entries) => {
+      entries.forEach((entry) => {
+        setDimensions(entry.contentRect);
+      });
+    });
+    resizeObserver.observe(observeTarget);
+    return () => {
+      resizeObserver.unobserve(observeTarget);
+    };
+  }, [ref]);
+  return dimensions;
+};
+
+function usePrevious(value) {
```

Comment:
Seems clearer to just put this in the main `Tree` function; i.e. could just put the previous variable defined as useRef({}) in there and update in the useEffect for the jsonData

Reply:
I think it's better to keep this logic, for storing the previous object, in its own function rather than cluttering the useEffect for jsonData.